runtime/compiler/env/j9method.cpp

```diff
@@ -9312,17 +9312,61 @@ TR_J9ByteCodeIlGenerator::packReferenceChainOffsets(TR::Node *node, std::vector<
 
 bool
 TR_ResolvedJ9Method::isFieldQType(int32_t cpIndex)
+   {
+   J9ROMFieldRef *ref = (J9ROMFieldRef *) (&romCPBase()[cpIndex]);
+   J9ROMNameAndSignature *nameAndSignature = J9ROMFIELDREF_NAMEANDSIGNATURE(ref);
+   J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSignature);
+
+   J9VMThread *vmThread = fej9()->vmThread();
+   return vmThread->javaVM->internalVMFunctions->isNameOrSignatureQtype(signature);
+   }
+
+bool
+TR_ResolvedJ9Method::isFieldNullRestricted(TR::Compilation *comp, int32_t cpIndex, bool isStatic, bool isStore)
    {
    if (!TR::Compiler->om.areFlattenableValueTypesEnabled() ||
       (-1 == cpIndex))
       return false;
 
+   if (TR::Compiler->om.isQDescriptorForValueTypesSupported())
+      {
+      if (isFieldQType(cpIndex)) // Temporary until javac supports NullRestricted attribute
+         return true;
+      }
+
+   IDATA fieldOffset = -1;
+   void *staticAddress = NULL;
    J9VMThread *vmThread = fej9()->vmThread();
-   J9ROMFieldRef *ref = (J9ROMFieldRef *) (&romCPBase()[cpIndex]);
-   J9ROMNameAndSignature *nameAndSignature = J9ROMFIELDREF_NAMEANDSIGNATURE(ref);
-   J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSignature);
+   J9ROMFieldShape *fieldShape = NULL;
 
-   return vmThread->javaVM->internalVMFunctions->isNameOrSignatureQtype(signature);
+   {
+   TR::VMAccessCriticalSection isFieldNullRestricted(fej9());
+   if (isStatic)
+      {
+      staticAddress = jitCTResolveStaticFieldRefWithMethod(vmThread, ramMethod(), cpIndex, isStore, &fieldShape);
+      }
+   else
+      {
+      fieldOffset = jitCTResolveInstanceFieldRefWithMethod(vmThread, ramMethod(), cpIndex, isStore, &fieldShape);
+      }
+   }
+
+   if (isStatic)
+      {
+      if (!staticAddress)
+         {
+         comp->failCompilation<TR::CompilationException>("jitCTResolveStaticFieldRefWithMethod failed");
+         }
+      }
+   else
+      {
+      if (fieldOffset == -1)
+         {
+         comp->failCompilation<TR::CompilationException>("jitCTResolveInstanceFieldRefWithMethod failed");
+         }
+      }
+
+   return vmThread->javaVM->internalVMFunctions->isFieldNullRestricted(fieldShape);
```

Comment:
This might be a little easier to read if it checked for failure conditions inside the first `if (isStatic)` above.  I know you don't want to fail the compilation inside there, because it's holding VM access, but maybe the first `if` could set a `bool failCompilation` variable and a `compilationFailureMsg` variable.  Then this second `if` could be replaced with:

```
if (failCompilation)
   {
   comp->failCompilation<TR::CompilationException>(compilationFailureMsg);
   }
```

Reply:
Addressed in 32f6437